### PR TITLE
Refactor code and add Build API support

### PR DESCRIPTION
**Pull request by jlebon**
- Use [pykwalify](http://pykwalify.readthedocs.io/) to validate YAML. It's not perfect, but it goes a long way towards strengthening handling of malformed YAML. There is some custom handling required to get the more complex keys checked properly.
- Refactor the testrunner. This is mostly in preparation for sanely adding support for the Build API. Though it adds a lot of nice features, such as:
  - Unified codepaths for containerized and virtualized tests.
  - Stronger timeout handling and environment cleanups.
  - Nicer logs, with time taken per step and separation between package install and testing.
- Add support for the Build API (#11).

### CODE REVIEW

File: sample.redhat-ci.yml [hunk 1]
```diff
@@ -83,6 +83,28 @@ env:
...
+build:
+    # OPTIONAL
+    # Customize the configuration step.
+    config-opts: >
+        --prefix=/usr
+        --libdir=/usr/lib64
+        CFLAGS='-ggdb -O0'
+    # OPTIONAL
+    # Customize the build step.
+    build-opts: -j4
```
Comment (cgwalters) [inline sample.redhat-ci.yml]: I think parallelization is something that should be injected by the host system, not chosen per user.  We should just default to `-j $(getconf _NPROCESSORS_ONLN)`.

File: sample.redhat-ci.yml [hunk 2]
```diff
@@ -83,6 +83,28 @@ env:
...
+
+# You may customize the build further by providing a dict
+# instead.
+build:
+    # OPTIONAL
+    # Customize the configuration step.
+    config-opts: >
+        --prefix=/usr
+        --libdir=/usr/lib64
+        CFLAGS='-ggdb -O0'
```
Comment (cgwalters) [inline sample.redhat-ci.yml]: I'd use something like `--enable-systemd CFLAGS='-fsanitize=undefined'`.  The options you have there are really defaults.

File: testrunner [hunk 3]
```diff
@@ -266,35 +266,87 @@ install_packages() {
...
+        touch $state/build.sh
+        if [ ! -f $checkout/configure ]; then
+            if [ -f $checkout/autogen.sh ]; then
+                echo "NOCONFIGURE=1 ./autogen.sh" >> $state/build.sh
+            elif [ -f $checkout/autogen ]; then
+                echo "NOCONFIGURE=1 ./autogen" >> $state/build.sh
+            fi
+        fi
+
+        echo "./configure $config_opts" >> $state/build.sh
```
Comment (cgwalters) [inline testrunner:333]: I'd consider doing separate `srcdir != builddir` by default, like GContinuous does.  It makes it easier in the future to parallelize builds of the same source.  (debug/optimized etc).

**Review by cgwalters — COMMENTED** [review]
I only scanned this.  Seems sane though.

File: sample.redhat-ci.yml (hunk 2 above)
Comment (jlebon) [reply sample.redhat-ci.yml]: Fair enough, no point in showcasing defaults.

File: sample.redhat-ci.yml (hunk 1 above)
Comment (jlebon) [reply sample.redhat-ci.yml]: Yeah agreed. Thanks for the `getconf` pointer.

File: testrunner (hunk 3 above)
Comment (jlebon) [reply testrunner:333]: That makes sense, though I'll leave it for now until it's needed.

**Comment by jlebon** [comment]
:arrow_up: Added more interesting sample opts and make of `getconf`.

File: testrunner [hunk 4]
```diff
@@ -255,6 +255,17 @@ install_packages() {
         mgr=dnf
     fi
 
+    # This is hacky and sad. Preemptively try to refresh
+    # yum/dnf cache to work around flaky distro infras.
```
Comment (cgwalters) [inline testrunner:259]: rpm-md metadata + fedora's model of deleting packages is fundamentally racy.  The only fix is to have a local repo that has multiple copies of both metadata and packages.

**Comment by cgwalters** [comment]
Looks OK.

**Pull request closed without merging**